File: dexfile/src/main/java/com/github/netomi/bat/dexfile/value/EncodedValue.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.value;

import com.github.netomi.bat.dexfile.DexFile;
import com.github.netomi.bat.dexfile.io.DexDataInput;
import com.github.netomi.bat.dexfile.io.DexDataOutput;
import com.github.netomi.bat.dexfile.io.DexFormatException;
import com.github.netomi.bat.dexfile.visitor.EncodedValueVisitor;
import com.github.netomi.bat.util.Primitives;

public abstract class EncodedValue
{
    // Value types.

    public static final int VALUE_BYTE          = 0x00;
    public static final int VALUE_SHORT         = 0x02;
    public static final int VALUE_CHAR          = 0x03;
    public static final int VALUE_INT           = 0x04;
    public static final int VALUE_LONG          = 0x06;
    public static final int VALUE_FLOAT         = 0x10;
    public static final int VALUE_DOUBLE        = 0x11;
    public static final int VALUE_METHOD_TYPE   = 0x15;
    public static final int VALUE_METHOD_HANDLE = 0x16;
    public static final int VALUE_STRING        = 0x17;
    public static final int VALUE_TYPE          = 0x18;
    public static final int VALUE_FIELD         = 0x19;
    public static final int VALUE_METHOD        = 0x1a;
    public static final int VALUE_ENUM          = 0x1b;
    public static final int VALUE_ARRAY         = 0x1c;
    public static final int VALUE_ANNOTATION    = 0x1d;
    public static final int VALUE_NULL          = 0x1e;
    public static final int VALUE_BOOLEAN       = 0x1f;

    public abstract int getValueType();

    public abstract void readValue(DexDataInput input, int valueArg);

    protected final int writeType(DexDataOutput output, int valueArg) {
        int typeAndArg = (valueArg << 5) | getValueType();
        output.writeUnsignedByte((byte) typeAndArg);
        return valueArg;
    }

    protected abstract int writeType(DexDataOutput output);

    public final void write(DexDataOutput output) {
        int valueArg = writeType(output);
        writeValue(output, valueArg);
    }

    public abstract void writeValue(DexDataOutput output, int valueArg);

    public abstract void accept(DexFile dexFile, EncodedValueVisitor visitor);

    @Override
    public abstract int hashCode();

    @Override
    public abstract boolean equals(Object obj);

    public static EncodedValue read(DexDataInput input) {
        int typeAndArg = input.readUnsignedByte();

        int valueArg = typeAndArg >>> 5;
        int valueType = typeAndArg & 0x1f;

        EncodedValue encodedValue = create(valueType);

        encodedValue.readValue(input, valueArg);
        return encodedValue;
    }

    private static EncodedValue create(int valueType) {
        switch (valueType) {
            case VALUE_BYTE:          return new EncodedByteValue();
            case VALUE_SHORT:         return new EncodedShortValue();
            case VALUE_CHAR:          return new EncodedCharValue();
            case VALUE_INT:           return new EncodedIntValue();
            case VALUE_LONG:          return new EncodedLongValue();
            case VALUE_FLOAT:         return new EncodedFloatValue();
            case VALUE_DOUBLE:        return new EncodedDoubleValue();
            case VALUE_METHOD_TYPE:   return new EncodedMethodTypeValue();
            case VALUE_METHOD_HANDLE: return new EncodedMethodHandleValue();
            case VALUE_STRING:        return new EncodedStringValue();
            case VALUE_TYPE:          return new EncodedTypeValue();
            case VALUE_FIELD:         return new EncodedFieldValue();
            case VALUE_METHOD:        return new EncodedMethodValue();
            case VALUE_ENUM:          return new EncodedEnumValue();
            case VALUE_ARRAY:         return new EncodedArrayValue();
            case VALUE_ANNOTATION:    return new EncodedAnnotationValue();
            case VALUE_NULL:          return EncodedNullValue.instance();
            case VALUE_BOOLEAN:       return new EncodedBooleanValue();
            default: throw new DexFormatException("Unexpected EncodedValue type: " + Primitives.toHexString((short) valueType));
        }
    }

    public static EncodedAnnotationValue readAnnotationValue(DexDataInput input) {
        EncodedAnnotationValue annotation = new EncodedAnnotationValue();
        annotation.readValue(input, 0);
        return annotation;
    }

    static int requiredBytesForSignedInt(int value) {
        return (value << 24 >> 24 == value) ? 1 :
               (value << 16 >> 16 == value) ? 2 :
               (value << 8  >>  8 == value) ? 3 : 4;
    }

    static int requiredBytesForSignedLong(long value) {
        return (value << 56 >> 56 == value) ? 1 :
               (value << 48 >> 48 == value) ? 2 :
               (value << 40 >> 40 == value) ? 3 :
               (value << 32 >> 32 == value) ? 4 :
               (value << 24 >> 24 == value) ? 5 :
               (value << 16 >> 16 == value) ? 6 :
               (value << 8  >>  8 == value) ? 7 : 8;
    }

    static int requiredBytesForSignedShort(short value) {
        return (value << 24 >> 24 == value) ? 1 : 2;
    }

    static int requiredBytesForUnsignedChar(char value) {
        return (value << 24 >>> 24 == value) ? 1 : 2;
    }

    static int requiredBytesForUnsignedInt(int value) {
        return (value << 24 >>> 24 == value) ? 1 :
               (value << 16 >>> 16 == value) ? 2 :
               (value << 8  >>>  8 == value) ? 3 : 4;
    }

    static int requiredBytesForFloat(float value) {
        int bits = Float.floatToIntBits(value);

        return (bits <<  8 == 0) ? 1 :
               (bits << 16 == 0) ? 2 :
               (bits << 24 == 0) ? 3 : 4;
    }

    static int requiredBytesForDouble(double value) {
        long bits = Double.doubleToLongBits(value);

        return (bits <<  8 == 0) ? 1 :
               (bits << 16 == 0) ? 2 :
               (bits << 24 == 0) ? 3 :
               (bits << 32 == 0) ? 4 :
               (bits << 40 == 0) ? 5 :
               (bits << 48 == 0) ? 6 :
               (bits << 56 == 0) ? 7 : 8;
    }
}

File: dexfile/src/main/java/com/github/netomi/bat/dexfile/value/EncodedValue.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.value

import com.github.netomi.bat.dexfile.*
import com.github.netomi.bat.dexfile.io.DexDataInput
import com.github.netomi.bat.dexfile.io.DexDataOutput
import com.github.netomi.bat.dexfile.io.DexFormatException
import com.github.netomi.bat.dexfile.value.visitor.EncodedValueVisitor
import com.github.netomi.bat.dexfile.visitor.ReferencedIDVisitor
import com.github.netomi.bat.util.toHexStringWithPrefix

/**
 * An abstract base class representing an encoded value inside a dex file.
 *
 * @see <a href="https://source.android.com/devices/tech/dalvik/dex-format#encoding">value encoding @ dex format</a>
 */
abstract class EncodedValue {
    abstract val valueType: EncodedValueType

    internal abstract fun readValue(input: DexDataInput, valueArg: Int)

    protected fun writeType(output: DexDataOutput, valueArg: Int): Int {
        val typeAndArg = valueArg shl 5 or valueType.value
        output.writeUnsignedByte(typeAndArg.toShort())
        return valueArg
    }

    protected abstract fun writeType(output: DexDataOutput): Int

    internal fun write(output: DexDataOutput) {
        val valueArg = writeType(output)
        writeValue(output, valueArg)
    }

    internal abstract fun writeValue(output: DexDataOutput, valueArg: Int)

    abstract fun accept(dexFile: DexFile, visitor: EncodedValueVisitor)

    internal abstract fun referencedIDsAccept(dexFile: DexFile, visitor: ReferencedIDVisitor)

    abstract override fun hashCode(): Int

    abstract override fun equals(other: Any?): Boolean

    companion object {
        fun read(input: DexDataInput): EncodedValue {
            val typeAndArg = input.readUnsignedByte().toInt()
            val valueArg   = typeAndArg ushr 5
            val valueType  = typeAndArg and 0x1f
            val encodedValue = create(valueType)
            encodedValue.readValue(input, valueArg)
            return encodedValue
        }

        private fun create(valueType: Int): EncodedValue {
            return EncodedValueType.of(valueType)?.supplier?.invoke()
                ?: throw DexFormatException("unexpected encoded value type: ${toHexStringWithPrefix(valueType.toByte())}")
        }

        fun readAnnotationValue(input: DexDataInput): EncodedAnnotationValue {
            val annotation = EncodedAnnotationValue.empty()
            annotation.readValue(input, 0)
            return annotation
        }

        fun requiredBytesForSignedInt(value: Int): Int {
            return when (value) {
                value shl 24 shr 24 -> 1
                value shl 16 shr 16 -> 2
                value shl  8 shr  8 -> 3
                else                -> 4
            }
        }

        fun requiredBytesForSignedLong(value: Long): Int {
            return when (value) {
                value shl 56 shr 56 -> 1
                value shl 48 shr 48 -> 2
                value shl 40 shr 40 -> 3
                value shl 32 shr 32 -> 4
                value shl 24 shr 24 -> 5
                value shl 16 shr 16 -> 6
                value shl  8 shr  8 -> 7
                else                -> 8
            }
        }

        fun requiredBytesForSignedShort(value: Short): Int {
            val v = value.toInt()
            return if (v shl 24 shr 24 == v) 1 else 2
        }

        fun requiredBytesForUnsignedChar(value: Char): Int {
            return if (value.code shl 24 ushr 24 == value.code) 1 else 2
        }

        fun requiredBytesForUnsignedInt(value: Int): Int {
            return when (value) {
                value shl 24 ushr 24 -> 1
                value shl 16 ushr 16 -> 2
                value shl  8 ushr  8 -> 3
                else                 -> 4
            }
        }

        fun requiredBytesForFloat(value: Float): Int {
            val bits = java.lang.Float.floatToIntBits(value)

            return      if (bits shl  8 == 0) 1
                   else if (bits shl 16 == 0) 2
                   else if (bits shl 24 == 0) 3
                   else                       4
        }

        fun requiredBytesForDouble(value: Double): Int {
            val bits = java.lang.Double.doubleToLongBits(value)

            return      if (bits shl  8 == 0L) 1
                   else if (bits shl 16 == 0L) 2
                   else if (bits shl 24 == 0L) 3
                   else if (bits shl 32 == 0L) 4
                   else if (bits shl 40 == 0L) 5
                   else if (bits shl 48 == 0L) 6
                   else if (bits shl 56 == 0L) 7
                   else                        8
        }
    }
}

enum class EncodedValueType constructor(val value: Int, val supplier: () -> EncodedValue) {
    BYTE          (VALUE_BYTE,          { EncodedByteValue.empty() }),
    SHORT         (VALUE_SHORT,         { EncodedShortValue.empty() }),
    CHAR          (VALUE_CHAR,          { EncodedCharValue.empty() }),
    INT           (VALUE_INT,           { EncodedIntValue.empty() }),
    LONG          (VALUE_LONG,          { EncodedLongValue.empty() }),
    FLOAT         (VALUE_FLOAT,         { EncodedFloatValue.empty() }),
    DOUBLE        (VALUE_DOUBLE,        { EncodedDoubleValue.empty() }),
    METHOD_TYPE   (VALUE_METHOD_TYPE,   { EncodedMethodTypeValue.empty() }),
    METHOD_HANDLE (VALUE_METHOD_HANDLE, { EncodedMethodHandleValue.empty() }),
    STRING        (VALUE_STRING,        { EncodedStringValue.empty() }),
    TYPE          (VALUE_TYPE,          { EncodedTypeValue.empty() }),
    FIELD         (VALUE_FIELD,         { EncodedFieldValue.empty() }),
    METHOD        (VALUE_METHOD,        { EncodedMethodValue.empty() }),
    ENUM          (VALUE_ENUM,          { EncodedEnumValue.empty() }),
    ARRAY         (VALUE_ARRAY,         { EncodedArrayValue.empty() }),
    ANNOTATION    (VALUE_ANNOTATION,    { EncodedAnnotationValue.empty() }),
    NULL          (VALUE_NULL,          { EncodedNullValue }),
    BOOLEAN       (VALUE_BOOLEAN,       { EncodedBooleanValue.empty() });

    companion object {
        private val valueToTypeMap: Map<Int, EncodedValueType> by lazy {
            values().associateBy { it.value }
        }

        fun of(valueType: Int) : EncodedValueType? {
            return valueToTypeMap[valueType]
        }
    }
}